astropy/stats/circstats.py

```diff
@@ -0,0 +1,448 @@
+    The definition used here differs from the one in scipy.stats.circvar.
+    Precisely, Scipy circvar uses an approximation based on the limit of small
+    angles which approaches the linear variance.
+    """
+
+    return 1.0 - _length(data, w, 1, 0.0, axis)
+
+
+def circmoment(data, w=None, p=1.0, centered=False, axis=None):
+    """ Computes the ``p``-th trigonometric circular moment for an array
+    of circular data.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Array of circular (directional) data.
+    w : numpy.ndarray, optional
+        In case of grouped data, the i-th element of ``w`` represents a
+        weighting factor for each group such that sum(w,axis) equals the number
+        of observations. See [1], remark 1.4, page 22, for detailed
+        explanation.
+    p : float, optional
+        Order of the circular moment.
+    centered : Boolean, optinal
+        If ``True``, central circular moments are computed. Default value is
+        ``False``.
+    axis : int, optional
+        Axis along which circular moments are computed. The default is to
+        compute the circular moment of the flattened array.
+
+    Returns
+    -------
+    circmoment: Quantity
+        The first and second elements correspond to the direction and length of
+        the ``p``-th circular moment, respectively.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from astropy.stats import circmoment
+    >>> from astropy import units as u
+    >>> data = np.array([51, 67, 40, 109, 31, 358])*u.deg
+    >>> circmoment(data, p=2) # doctest: +FLOAT_CMP
+    (<Quantity 90.99263082432564 deg>, <Quantity 0.48004283892950717>)
+
+    References
+    ----------
+    .. [1] S. R. Jammalamadaka, A. SenGupta. "Topics in Circular Statistics".
+       Series on Multivariate Analysis, Vol. 5, 2001.
+    .. [2] C. Agostinelli, U. Lund. "Circular Statistics from 'Topics in
+       Circular Statistics (2001)'". 2015.
+       <https://cran.r-project.org/web/packages/CircStats/CircStats.pdf>
+    """
+    phi = 0.0
+    if centered:
+        phi = circmean(data, w, axis)
+
+    return _angle(data, w, p, phi, axis), _length(data, w, p, phi, axis)
+
+
+def circcorrcoef(alpha, beta, w_alpha=None, w_beta=None, ax_alpha=None,
+                 ax_beta=None):
+    """ Computes the circular correlation coefficient between two array of
+    circular data.
+
+    Parameters
+    ----------
+    alpha : numpy.ndarray
+        Array of circular (directional) data.
+    beta : numpy.ndarray
+        Array of circular (directional) data.
+    w_alpha : numpy.ndarray, optional
+        In case of grouped data, the i-th element of ``w_alpha`` represents a
+        weighting factor for each group such that sum(w,axis) equals the number
+        of observations. See [1], remark 1.4, page 22, for detailed
+        explanation.
+    w_beta : numpy.ndarray, optional
+        See description of ``w_alpha``.
+    ax_alpha : int, optional
```

Comment:
Should we really do separate axes for the two arrays? Maybe it is cleaner to just give one axis and let the caller organise things? I wonder mostly since broadcasting along the other axes could make things somewhat surprising.